mono/mini/interp/transform.c

```diff
@@ -1484,6 +1484,61 @@ interp_handle_intrinsics (TransformData *td, MonoMethod *target_method, MonoClas
 		if (!strcmp (tm, "MemoryBarrier") && csignature->param_count == 0)
 			*op = MINT_MONO_MEMORY_BARRIER;
 #endif
+	} else if (in_corlib &&
+			!strcmp (klass_name_space, "System.Runtime.CompilerServices") &&
+			!strcmp (klass_name, "JitHelpers") &&
+			(!strcmp (tm, "EnumEquals") || !strcmp (tm, "EnumCompareTo"))) {
+		MonoGenericContext *ctx = mono_method_get_context (target_method);
+		g_assert (ctx);
+		g_assert (ctx->method_inst);
+		g_assert (ctx->method_inst->type_argc == 1);
+		g_assert (csignature->param_count == 2);
+
+		MonoType *t = ctx->method_inst->type_argv [0];
+		t = mini_get_underlying_type (t);
+
+		gboolean is_i8 = (t->type == MONO_TYPE_I8 || t->type == MONO_TYPE_U8);
+		gboolean is_unsigned = (t->type == MONO_TYPE_U1 || t->type == MONO_TYPE_U2 || t->type == MONO_TYPE_U4 || t->type == MONO_TYPE_U8 || t->type == MONO_TYPE_U);
+
+		gboolean is_compareto = strcmp (tm, "EnumCompareTo") == 0;
+		if (is_compareto) {
+			int offseta, offsetb;
+			offseta = create_interp_local (td, t);
+			offsetb = create_interp_local (td, t);
+
+			// Save arguments
+			store_local_general (td, offsetb, t);
+			store_local_general (td, offseta, t);
+			// (a > b)
+			load_local_general (td, offseta, t);
+			load_local_general (td, offsetb, t);
+			if (is_unsigned)
+				interp_add_ins (td, is_i8 ? MINT_CGT_UN_I8 : MINT_CGT_UN_I4);
+			else
+				interp_add_ins (td, is_i8 ? MINT_CGT_I8 : MINT_CGT_I4);
+			td->sp --;
+			SET_SIMPLE_TYPE (td->sp - 1, STACK_TYPE_I4);
+			// (a < b)
+			load_local_general (td, offseta, t);
+			load_local_general (td, offsetb, t);
+			if (is_unsigned)
+				interp_add_ins (td, is_i8 ? MINT_CLT_UN_I8 : MINT_CLT_UN_I4);
+			else
+				interp_add_ins (td, is_i8 ? MINT_CLT_I8 : MINT_CLT_I4);
+			td->sp --;
+			SET_SIMPLE_TYPE (td->sp - 1, STACK_TYPE_I4);
+			// (a > b) - (a < b)
+			interp_add_ins (td, MINT_SUB_I4);
+			td->sp --;
+			td->ip += 5;
+			return TRUE;
+		} else {
+			if (is_i8) {
+				*op = MINT_CEQ_I8;
+			} else {
+				*op = MINT_CEQ_I4;
+			}
```

Comment:
There are also float/double/char/[s]byte/[u]short enums. Not sure if the code will work correctly for them. The netcore configuration has tests for them in System.Runtime.Tests.

Reply:
@filipnavara I doubt float/double are valid enums. On the interpreter stack we cast i1 and i2 to i4 and store it as such on the stack. So CEQ_I4 is ok.

Reply:
We had a long discussion about the validity of float/double enums at https://github.com/dotnet/corefx/issues/36887 and linked issues. They are not valid according to current ECMA specification but previous revisions didn't state that explicitly and the CLR partially supports them. Mono now handles some subset of the operations on float/double enums in the JIT so I believe the interpreter should match the results. I am fine with leaving that out of this PR but I wanted to mention it nevertheless.

Reply:
[U]IntPtr is still valid according to the current ECMA specification though.

Reply:
This pr brings the same support as with jit. We can add later missing pieces.

Reply:
JIT supports native int sized enums (eg. https://github.com/mono/mono/blob/d540aee31f71edb901fe6239f04bb29ee6e3c0d6/mono/metadata/icall.c#L3938-L3944). I am fine leaving it to separate PR, especially since these enums cannot be created with regular C#.

Reply:
You handle `MONO_TYPE_U` in `is_unsigned` but it's not handled in `is_i8` for 64-bit targets.